src/shared/utilities/tsUtils.ts

```diff
@@ -31,3 +31,13 @@ type Expand<T> = T extends infer O ? { [K in keyof O]+?: O[K] } : never
  *
  */
 export type ExpandWithObject<T> = Expand<T> extends Record<string, unknown> ? Expand<T> : never
+
+// Special class to embed a hidden 'key' into a type
+abstract class NominalKey<K> {
```

Comment:
Why this works:
Classes are a special case in TypeScript in that the type-checker will check for _all_ fields, even those marked private/protected. We can then create intersections with arbitrary _T_ types, but on the outside these intersections still look like _T_. But now if you tried to assign _T_ to `T & NominalKey<K>` it will not work because _T_ is missing `__type__: K`. Since the class is abstract, we don't need to assign anything to the field, making the type itself ephemeral. 

Basically:
`T extends (T & NominalKey<K>)` <-- FALSE
`(T & NominalKey<K>) extends T` <-- TRUE
The intersection type is always a valid sub-type of _T_, but the inverse is not true.